### PR TITLE
[VirtualMachine] Zero copy in set_input when input is DLTensor

### PR DESCRIPTION
**Pull request by vvchernov**
I observed that VirtualMachine::SetInputTensorWithIndex(...) method has discrepancy between description (also see description for VirtualMachine::SetInput(...) which assumes zero copy if possible and uses the method) and implementation. It always create new NDArray and copies data to it if source input is DLTensor even if devices are the same. It reduces performance of multiple input models due to excess copying. The PR fixes this issue.

Note: I have a remark about current design. VirtualMachine has only `set_input` python method, the same method is used inside `run` and `invoke` methods with input args. But there is no `set_input_zero_copy`. In description I obsrved that `set_input` tries to not use copying if possible. Theoretically we can have problem if `set_input` is used, input tensors are released after that and when `run` or `invoke` are launched. As I know GraphExecutor does not have such problem.

### CODE REVIEW

**Review by jwfromm — APPROVED** [review]
I like this change a lot and your comments are excellent.

**Review by AndrewZhaoLuo — COMMENTED** [review]
Hmm, I'm not sure about this change, it seems like a major change in invariants. I would rather you make a new method like "set_input_zero_copy" and expose that to the user to use.

**Review by AndrewZhaoLuo — CHANGES_REQUESTED** [review]
Gonna block this until get another set of opinions on this. @mbs-octoml @altanh ?

**Comment by vvchernov** [comment]
Hello @AndrewZhaoLuo! As you can see my note to the PR I said about the same, but currently `set_input` method of VirtualMachine already does zero copy for NDArray input. It means that if I add `set_input_zero_copy` I should change `set_input` to method which always copies external input to internal one (and performance for people who used it in their code is reduced). No problem, I can do it for me it is more reasonable approach. What do you think @jwfromm?

File: src/runtime/vm/vm.cc [hunk 1]
```diff
@@ -303,13 +303,12 @@ void VirtualMachine::SetInputTensorWithIndex(std::vector<ObjectRef>& tensors,
   if (inp_tensor.type_code() == kTVMDLTensorHandle) {
     // Automatically convert input DLTensors to NDArray
     DLTensor* tensor = inp_tensor;
-    std::vector<int64_t> shape;
-    for (int64_t i = 0; i < tensor->ndim; i++) {
-      shape.push_back(tensor->shape[i]);
+    if (dev.device_type == tensor->device.device_type &&
```
Comment (mbs-octoml) [inline src/runtime/vm/vm.cc:306]: Thank you, this is a great change.

Could you update vm.py's set_input doc string to clearly state the by-copy vs by-ref semantics? Thanks.

**Review by mbs-octoml — APPROVED** [review]
LGTM. Just a comment nit. Thanks, copy overhead has been troubling me lately so I'm glad you're ahead of me.

**Review by AndrewZhaoLuo — APPROVED** [review]

**Comment by AndrewZhaoLuo** [comment]
> Hello @AndrewZhaoLuo! As you can see my note to the PR I said about the same, but currently `set_input` method of VirtualMachine already does zero copy for NDArray input. It means that if I add `set_input_zero_copy` I should change `set_input` to method which always copies external input to internal one (and performance for people who used it in their code is reduced). No problem, I can do it for me it is more reasonable approach. What do you think @jwfromm?

After talking to MBS, your change does in fact match the intended semantics better

**Comment by AndrewZhaoLuo** [comment]
Just please cover the nit

File: src/runtime/vm/vm.cc (hunk 1 above)
Comment (vvchernov) [reply src/runtime/vm/vm.cc:313]: Hello @mbs-octoml! I've added description and check device id for NDArray. But it looks like internal mechanism of copying does not take into account device id. Please see my changes

Comment (mbs-octoml) [reply src/runtime/vm/vm.cc:313]: Thanks for that. Yeah, the codebase is not at all 'device_id clean' and will require an audit to find all the places it is either ignored or defaulted to '0'. One step at a time.

**Pull request merged**